PC/_wmimodule.cpp

```diff
@@ -0,0 +1,310 @@
+                        !WriteFile(data->writePipe, propStr, cbStr2, &written, NULL) ||
+                        !WriteFile(data->writePipe, (LPVOID)L"\0", 2, &written, NULL)
+                    ) {
+                        hr = HRESULT_FROM_WIN32(GetLastError());
+                    }
+                }
+                VariantClear(&propValue);
+                SysFreeString(propName);
+            }
+        }
+        value->EndEnumeration();
+        value->Release();
+    }
+
+    if (bstrQuery) {
+        SysFreeString(bstrQuery);
+    }
+    if (enumerator) {
+        enumerator->Release();
+    }
+    if (services) {
+        services->Release();
+    }
+    if (locator) {
+        locator->Release();
+    }
+    CoUninitialize();
+    CloseHandle(data->writePipe);
+    return (DWORD)hr;
+}
+
+
+/*[clinic input]
+_wmi.exec_query
+
+    query: unicode
+
+Runs a WMI query against the local machine.
+
+This returns a single string with 'name=value' pairs in a flat array separated
+by null characters.
+[clinic start generated code]*/
+
+static PyObject *
+_wmi_exec_query_impl(PyObject *module, PyObject *query)
+/*[clinic end generated code: output=a62303d5bb5e003f input=48d2d0a1e1a7e3c2]*/
+
+/*[clinic end generated code]*/
+{
+    PyObject *result = NULL;
+    HANDLE hThread = NULL;
+    int err = 0;
+    WCHAR buffer[8192];
+    DWORD offset = 0;
+    DWORD bytesRead;
+    struct _query_data data = {0};
+
+    if (PySys_Audit("_wmi.exec_query", "O", query) < 0) {
+        return NULL;
+    }
+
+    data.query = PyUnicode_AsWideCharString(query, NULL);
+    if (!data.query) {
+        return NULL;
+    }
+
+    if (0 != _wcsnicmp(data.query, L"select ", 7)) {
+        PyMem_Free((void *)data.query);
+        PyErr_SetString(PyExc_ValueError, "only SELECT queries are supported");
+        return NULL;
+    }
+
+    Py_BEGIN_ALLOW_THREADS
+
+    if (!CreatePipe(&data.readPipe, &data.writePipe, NULL, 0)) {
+        err = GetLastError();
+    } else {
+        hThread = CreateThread(NULL, 0, _query_thread, (LPVOID*)&data, 0, NULL);
+        if (!hThread) {
+            err = GetLastError();
+            // Normally the thread proc closes this handle, but since we never started
+            // we need to close it here.
+            CloseHandle(data.writePipe);
+        }
+    }
+
+    while (!err) {
+        if (ReadFile(
+            data.readPipe,
+            (LPVOID)&buffer[offset / sizeof(buffer[0])],
+            sizeof(buffer) - offset,
+            &bytesRead,
+            NULL
+        )) {
+            offset += bytesRead;
+            if (offset >= sizeof(buffer)) {
+                err = ERROR_MORE_DATA;
+            }
+        } else {
+            err = GetLastError();
+        }
+    }
+
+    if (data.readPipe) {
+        CloseHandle(data.readPipe);
+    }
+
+    // Allow the thread some time to clean up
+    switch (WaitForSingleObject(hThread, 1000)) {
```

Comment:
Maybe make the timeout an optional parameter with a default value of 1 second? See _winapi.WaitForSingleObject() (which has a mandatory parameter).

Reply:
This timeout is after all the work is done. In testing, it only exceeded 1ms once under normal behaviour, but if COM shutdown decides to stall then we don't want to be stuck here forever (and we need the thread exit code to avoid having to set up volatile shared variables).

Basically, this timeout doesn't include the actual work. A 10 second request will still take 10 seconds (it'll probably fill the buffer and abort that way).